soroban-env-host/src/budget/dimension.rs

```diff
@@ -0,0 +1,176 @@
+use super::model::{HostCostModel, MeteredCostComponent, ScaledU64};
+use crate::xdr::{ContractCostParams, ContractCostType, ScErrorCode, ScErrorType};
+use crate::HostError;
+use core::fmt::Debug;
+
+/// Helper types to annotate boolean function arguments
+pub(crate) struct IsCpu(pub(crate) bool);
+pub(crate) struct IsInternal(pub(crate) bool);
+
+#[derive(Clone, Default)]
+pub(crate) struct BudgetDimension {
+    /// A set of cost models that map input values (eg. event counts, object
+    /// sizes) from some CostType to whatever concrete resource type is being
+    /// tracked by this dimension (eg. cpu or memory). CostType enum values are
+    /// used as indexes into this vector, to make runtime lookups as cheap as
+    /// possible.
+    pub(crate) cost_models: Vec<MeteredCostComponent>,
+
+    /// The limit against-which the count is compared to decide if we're
+    /// over budget.
+    pub(crate) limit: u64,
+
+    /// Tracks the output value from individual cost models
+    pub(crate) counts: Vec<u64>,
+
+    /// Tracks the sum of _output_ values from the cost model, for purposes
+    /// of comparing to limit.
+    pub(crate) total_count: u64,
+
+    /// The internal limit tracks work done internally that is not exposed to the
+    /// external user -- it does not affect fees or decide the invocation outcome
+    /// in any way (no error due to exceeding the internal limit). It exists solely
+    /// for dos prevention. Such work include diagnostic logging, or work that
+    /// exists only for preflight.
+    pub(crate) internal_limit: u64,
```

Comment:
I thought we were going to introduce a separate budget object for this. I don't have a strong opinion on this (especially given that the current approach is already implemented) and I don't think we should change that, but I wonder if there is some fundamental issue that wouldn't allow us to use 2 regular budget structs and swap between them? If not, maybe add a todo/issue to revisit that later?

Reply:
There isn't a fundamental issue, i.e. we could have two budgets and switch them as you suggested. It's more of a design choice to keep the interface clean and with minimal disruption. 
- The full `Budget` structure contains many other fields (for tracking, debugging), which isn't necessary for the "internal" use case (detailed [above](https://github.com/stellar/rs-soroban-env/pull/1123#discussion_r1371164341)).
- The requirement for this work comes from the fact that we need some way to mitigate DOS from the diagnostic workflow, by specifying some limit (separate from normal budget) which we don't expect normal behaviors to exceed. So exposing that limit externally isn't strictly necessary, as long as the limit is not too large for DOS, and not too small to be easily tripped over. 
- Setting the internal limit same as the regular limit seem to be a good choice. We could also expose a scale factor as an internal host constant, but exposing it to the budget constructor feels like an overkill (in which case we may also consider exposing new network config params for it). 

Overall this is the most lightweight approach that appears sufficient towards achieving the goal, while keeping all external interfaces unchanged.

Reply:
Yeah, I agree this is serviceable for now. I still would feel a bit more comfortable if we had two distinct objects (so there is no concern that the internal logic is broken somewhere), but the usage seems limited enough to not be too concerned.